arrabiata/src/interpreter.rs

```diff
@@ -0,0 +1,962 @@
+//! The challenges will then be accumulated over time using the random coin used
+//! by the folding argument.
+//! The verifier circuit must be carefully implemented to ensure that all the
+//! messages that the prover would have sent before coining the random combiner
+//! for the constraints has been absorbed properly in the verifier circuit.
+//!
+//! Using this technique requires us a folding scheme that handles degree
+//! `5 + 1` constraints, as the challenge will be considered as a variable.
+//! The reader can refer to the folding library available in this monorepo for
+//! more contexts.
+//!
+//! ## Permutation argument
+//!
+//! Communication between rows must be done using a permutation argument. The
+//! argument we use will be a generalisation of the one used in the [PlonK
+//! paper](https://eprint.iacr.org/2019/953).
+//!
+//! The construction of the permutations will be done using the methods prefixed
+//! `save` and `load`. The index of the current row and the index of the
+//! time the value has been written will be used to generate the permutation on
+//! the fly.
+//!
+//! The permutation argument described in the PlonK paper is a kind of "inverse
+//! lookup" protocol, like Plookup. The polynomials are defined as follows:
+//!
+//! ```text
+//!            Can be seen as T[f(X)] = Χ
+//!          --------
+//!          |      |
+//! f'(X) = f(X) + β X + γ
+//!                      |--- Can be seen as the evaluation point.
+//!                         |
+//!                         |
+//! g'(X) = g(X) + β σ(X) + γ
+//!          |      |
+//!          --------
+//!          Can be seen as T[g(X)] = σ(X)
+//! ```
+//!
+//! And from this, we build an accumulator, like for Plookup.
+//! The accumulator requires to coin two challenges, β and γ, and it must be
+//! done after the commitments to the columns have been absorbed.
+//! The verifier at the next step will verify that the challenges have been
+//! correctly computed.
+//! In the implementation, the accumulator will be computed after the challenges
+//! and the commitments. Note that the accumulator must also be aggregated, and
+//! the aggregation must be performed by the verifier at the next step.
+//!
+//! The methods `save` and `load` will accept as arguments only a column that is
+//! included in the permutation argument. For instance, `save_poseidon_state`
+//! will only accept columns with index 3, 4 and 5, where the
+//! `load_poseidon_state` will only accepts columns with index 0, 1 and 2.
+//!
+//! The permutations values will be saved in public values, and will contain the
+//! index of the row. The permutation values will be encoded with a 32 bits
+//! value (u32) as we can suppose a negligible probability that a user will use
+//! more than 2^32 rows.
+//!
+//! The permutation argument also generates constraints that will be
+//! homogeneized with the gadget constraints.
+//!
+//! Note all rows might require to use the permutation argument. Therefore, a
```

Comment:
I understand that you want the save some computation if a written circuit does not copy constrains some columns ?
You do not need selectors to do that

Reply:
Discussed in a call: we talk about rows, and not columns.